nvi

```diff
@@ -0,0 +1,85 @@
+#!/bin/bash
+
+VERSION=1.0.0
+
+function node_version_installer() {
+  case "$1" in
+    --version)
+      echo "nvi $VERSION"
+      ;;
+    install)
+      if [[ -n "$2" ]]; then
+        # A second argument was given, so it should be a valid version string.
+        if [[ $2 =~ ^[0-9]+\.[0-9]+\.[0-9]+$ ]]; then
+          # If $2 is a seemingly valid version string, attempt to install it.
+          if [[ -n "$3" ]]; then
+            LOCAL_DIR=$3
+            BIN_DIR="$LOCAL_DIR/bin"
+          else
+            LOCAL_DIR="$HOME/.local"
+            BIN_DIR="$LOCAL_DIR/bin"
+          fi
+          mkdir --parents $BIN_DIR
+          # Download and extract the tar if it doesn't exist on disk.
+          if [[ ! -d "$LOCAL_DIR/node-v$2-linux-x64" ]]; then
+            # Fail if version doesn't exist.
+            NODE_DIST_LOOKUP=$(wget -qO - https://nodejs.org/dist/index.json | grep -Paoz "\"version\"\s*:\s*\"v$2\"")
+            [ -z "$NODE_DIST_LOOKUP" ] && >&2 echo "Couldn't find Node.js version v$2. Aborting." && exit 1
+            # Download it if it does.
+            NODE_TAR="https://nodejs.org/dist/v$2/node-v$2-linux-x64.tar.gz"
+            wget --output-document=$2.tar.gz --quiet $NODE_TAR
+            [ "$?" -ne 0 ] && echo "Couldn't download $NODE_TAR, but the version is valid. Possibly a network issue. Aborting." && exit 1
+            tar --extract --file $2.tar.gz --gzip --directory $LOCAL_DIR
+            rm $2.tar.gz
+          fi
+          cp $LOCAL_DIR/node-v$2-linux-x64/bin/node $BIN_DIR/node
+          ln --force --symbolic $(readlink --canonicalize $LOCAL_DIR/node-v$2-linux-x64/bin/npm) $BIN_DIR/npm
+        else
+          # Otherwise print some hints as to what constitutes a version string.
+          echo "Invalid version specified: $2. Correct format is [0-9]+.[0-9]+.[0-9]+"
+          exit 1
+        fi
+      else
+        if [[ -e "./package.json" ]]; then
+          # Found a package.json file, so read the version string from this.
+          NODE_VERSION=$(cat ./package.json | grep -Paoz '"engines"\s*:\s*{\s*"node"\s*:\s*"[^"]+"' | grep -Paoz '\d+\.\d+\.\d+')
```

Comment:
Currently this will only find a `NODE_VERSION` if it actually matches `^[0-9]+\.[0-9]+\.[0-9]+$`. This means we'll always have errors like

```text
Invalid version () in package.json. Correct format is [0-9]+.[0-9]+.[0-9]+
```

which is rather unhelpful. We need a more clever way to match specifically what's in `"node": "some string here"`.